File: 2307-replace-non-coprime-numbers-in-array/replace-non-coprime-numbers-in-array.java
class Solution {
    private long __gcd(long a, long b) {
        while (b != 0) {
            long rem = a % b;
            a = b;
            b = rem;
        }

        return a;
    }
    private long __lcm(long a, long b) {
        return (a * b) / __gcd(a, b);
    }
    public List<Integer> replaceNonCoprimes(int[] nums) {
        int n = nums.length;

        Deque<Integer> dq = new LinkedList<>();
        dq.addFirst(nums[0]);
        for (int i = 1; i < n; i++) {
            if (__gcd(Long.valueOf(dq.getLast()), nums[i]) != 1L) {
                long last = dq.removeLast();
                long lcm = __lcm(last, nums[i]);
                while (!dq.isEmpty()) {
                    if (__gcd(Long.valueOf(dq.getLast()), lcm) != 1L) {
                        lcm = __lcm(Long.valueOf(dq.removeLast()), lcm);
                    } else {
                        break;
                    }
                }
                dq.addLast((int)lcm);
            } else {
                dq.addLast(nums[i]);
            }
            // for (Integer ele: dq) {
            //     System.out.print(ele + " ");
            // }
            // System.out.println();
        }


        return new ArrayList<Integer>(dq);
    }
}

File: 2307-replace-non-coprime-numbers-in-array/replace-non-coprime-numbers-in-array.cpp
class Solution {
    using ll = long long;
    ll ____gcd(ll a, ll b) {
        while (b != 0) {
            ll rem = a % b;
            a = b;
            b = rem;
        }

        return a;
    }
    ll __lcm(ll& a, ll& b) {
        return (a * b) / __gcd(a, b);
    }
public:
    vector<int> replaceNonCoprimes(vector<int>& nums) {
        int n = nums.size();

        deque<ll> dq;
        dq.push_back(nums[0]);
        for (int i = 1; i < n; i++) {
            ll x = dq.back();
            ll y = nums[i];
            if (__gcd(x, y) != 1) {
                dq.pop_back();
                ll lcm = __lcm(x, y);
                while (!dq.empty()) {
                    ll ny = dq.back();
                    if (__gcd(ny, lcm) != 1) {
                        dq.pop_back();
                        lcm = __lcm(ny, lcm);
                    } else {
                        break;
                    }
                }
                dq.push_back(lcm);
            } else {
                dq.push_back(y);
            }
        }

        return vector<int>(dq.begin(), dq.end());
    }
};